book/src/snapshot-restore-feature.md

```diff
@@ -27,30 +30,96 @@ For details,  see the [CSI spec](https://github.com/container-storage-interface/
+* VolumeSnapshotSpec has a required Source field. Source may be either a PersistentVolumeClaimName (if dynamically provisioning a snapshot) or VolumeSnapshotContentName (if pre-provisioning a snapshot).
+* VolumeSnapshotContentSpec has a required Source field. This Source may be either a VolumeHandle (if dynamically provisioning a snapshot) or a SnapshotHandle (if pre-provisioning volume snapshots).
+* VolumeSnapshot contains a Status to indicate the current state of the volume snapshot. It has a field BoundVolumeSnapshotContentName to indicate the VolumeSnapshot object is bound to a VolumeSnapshotContent.
+* VolumeSnapshotContent contains a Status to indicate the current state of the volume snapshot content. It has a field SnapshotHandle to indicate that the VolumeSnapshotContent represents a snapshot on the storage system.
+
+### Controller Split
+
+* The CSI external-snapshotter sidecar is split into two controllers, a snapshot controller and a CSI external-snapshotter sidecar.
+
+The snapshot controller is deployed by the Kubernetes distributions and is responsible for watching the VolumeSnapshot CRD objects and manges the creation and deletion lifecycle of snapshots.
+
+The CSI external-snapshotter sidecar watches Kubernetes VolumeSnapshotContent CRD objects and triggers CreateSnapshot/DeleteSnapshot against a CSI endpoint.
+
+### Kubernetes Cluster Setup
+
+Volume snapshot is promoted to beta in Kubernetes 1.17 so the `VolumeSnapshotDataSource` feature gate is enabled by default.
+
+See the Deployment section of [Snapshot Controller](snapshot-controller.md) on how to set up the snapshot controller and CRDs.
+
+### Test Snapshot Feature
+
+To test snapshot Beta version, use the following [example yaml files](https://github.com/kubernetes-csi/external-snapshotter/tree/release-2.0/examples/kubernetes).
+
+Create a _StorageClass_:
+```
+kubectl create -f storageclass.yaml
+```
+
+Create a _PVC_:
+```
+kubectl create -f pvc.yaml
+```
+
+Create a _VolumeSnapshotClass_:
+```
+kubectl create -f snapshotclass.yaml
+```
 
-## Snapshot APIs
+Create a _VolumeSnapshot_:
+```
+kubectl create -f snapshot.yaml
+```
 
-Similar to the API for managing [Kubernetes Persistent Volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/), the Kubernetes Volume Snapshots introduce three new API objects for managing snapshots: `VolumeSnapshot`, `VolumeSnapshotContent`, and `VolumeSnapshotClass`. See [Kubernetes Snapshot documentation](https://kubernetes.io/docs/concepts/storage/volume-snapshots/) for more details.
+Create a _PVC_ from a _VolumeSnapshot_:
+```
+kuberctl create -f restore.yaml
+```
+
+#### PersistentVolumeClaim not Bound
+
+If a `PersistentVolumeClaim` is not bound, the attempt to create a volume snapshot from that `PersistentVolumeClaim` will fail. No retries will be attempted. An event will be logged to indicate that the `PersistentVolumeClaim` is not bound.
```

Comment:
How come retries are not attempted? This seems counter to the general reconciliation principles in Kubernetes controllers.

Reply:
This is because the exact time when the snapshot is taken is important for the applications.  We have decided to change this behavior and keep retrying when working on the create snapshot timeout issue, but that work is not complete yet.  So right now, this stays the same.  Will change after we have fixed that issue.

Reply:
Maybe add a reference to the issue saying that this behavior is planned to change

Reply:
Sure